### ISSUE TITLE
Add tracing support

### ISSUE THREAD

**Issue by khotyn** [alipay/sofa-mosn#246]
# Add tracing function

## General Design

As a sidecar, SOFAMosn's tracing requirement is much simpler than a general tracing framework such as Zipkin or Jaeger.

So in SOFAMosn, we can just provide a simple Span and Driver interface:

```go
type Span interface {
	SetOperation(operation string)

	SetTag(key string, value string)

	FinishSpan()

	InjectContext(requestHeaders map[string]string)

	SpawnChild(operationName string, startTime time.Time) Span
}

type Driver interface {
	Start(requestHeaders map[string]string, operationName string, startTime time.Time) Span
}
```

Other tracing framework/standard such as Zipkin or OpenTracing can act as an extension to `Span` and `Driver`.

### ~~Add request/response scoped context in connection layer~~

~~Since in SOFAMosn, protocols such as SOFARPC as decoded in connection layer, so it is obvious that we should create a Span either from the request header or from the void in the connection layer, but the problem is that connection's context is not per request/response, so it is not appropriate to use connection's context to store the Span.~~

~~**We should create a request/response scoped context from connection's context to store the Span, and the request/response scoped context should be the parent of stream's context, so in the stream layer, we can easily get the Span from stream's context and add more tags to it.**~~

### Add a SpanBuilder interface to the protocol layer

Since we want to limit the tracing related operations in the proxy layer, so a SpanBuilder interface should be add to the protocol layer, which is used to invoke by the proxy layer to build the span the add some protocol specific tags to the span.

The interface would be like this:

```go
type SpanBuilder interface {
	BuildSpan(context context.Context) Span
}
```

### Add a type to the listener to distinguish ingress from egress

In the egress scenario and the ingress scenario, there should be a different way to build the span. For example, in the ingress scenario, the tracing component should just accept the span from the downstream request, but in the egress scenario, after get the span id from the downstream request, the tracing component should spawn a child span.

So there should be a way the distinguish ingress from egress in the listener component, add and a type to the listener would be a reasonable way.

### Other considerations

* If there is no trace ID and span ID from downstream, SOFAMosn should generate a new trace ID and span ID then write the trace ID and span ID to the upstream request.
* HTTP/HTTP2 trace ID and span ID header should follow Envoy/Istio's behavior.
* A span ID factory should be designed to allow user to config how to generate the span ID. SOFATracer's span ID and Zipkin's span ID are quite different.
* Span is a request/response scoped struct, a pool should be considered to reuse the Spans to avoid GC.

## Configuration

Tracing configuration would be added under the server section of the configuration, such as:

```json
"tracing": {
	"sofarpc": {
		"driver": {
			"type": "SOFATracerLog",
			"spanIdType": "SOFATracerSpanId",
			"sample": 0 // Provide sampling options here
		}
	}
}
```

**Comment by khotyn** [comment]
@ldxdl @junxiong @Lingtaonju Pls have look at the tracing design for SOFAMosn. I already implement some feature that can generate SOFATracer log for SOFARPC for POC purpose, if you guys all agree with the design, I'll continue the work and submit a PR later.

**Comment by ldxdl** [comment]
@khotyn The solution is generally fine, I reserved tracing interface in types/trace.go already. The biggest problem is that the span‘s life cycle should be maintained in the proxy layer, not in connection layer. Currently there is a stream layer context generation mechanism in proxy, you should be familiar with proxy code first. It's best to talk to @taoyuanyuan for stream context details, it's a little complicated. The use of memory recycling is also ok to talk to him.

**Comment by nejisama** [comment]
fix #282
